feat(subscriber) expose server parts

hds opened the pull request:
The `ConsoleLayer` builder provides the user with a console layer and
a server, which is used to start the gRPC server.

However, it may be desireable to expose the instrumentation server together
with other services on the same Tonic router. This was requested
explicitly in #449 428.

Additionally, to add tests which make use of the instrumentation server
(as part of improving test coverage for #450), more flexibility is
needed than what is provided by the current API. Specifically we would
like to connect a client and server via an in memory channel, rather
than a TCP connection.

This change adds an additional method to `console_subscriber::Server`
called `into_parts` which allows the user to access the
`InstrumentServer` directly. The `Aggregator` is also returned and
must be set to run for at least as long as the instrument server. This
allows the aggregator to be spawned wherever the user wishes.

To facilitate the addition of functionality which would result in more
"parts" in the future, `into_parts` returns a non-exhaustive struct,
rather than a tuple of parts.

Closes: #428

hds committed:
feat(subscriber) expose server parts

The `ConsoleLayer` builder provides the user with a console layer and
a server, which is used to start the gRPC server.

However, it may be desireable to expose the instrumentation server together
with other services on the same Tonic router. This was requested
explicitly in #428.

Additionally, to add tests which make use of the instrumentation server
(as part of improving test coverage for #450), more flexibility is
needed than what is provided by the current API. Specifically we would
like to connect a client and server via an in memory channel, rather
than a TCP connection.

This change adds an additional method to `console_subscriber::Server`
called `into_parts` which allows the user to access the
`InstrumentServer` directly. A handle which controls the lifetime of the
`Aggregator` is also provided, as the user must ensure that the
aggregator lives at least as long as the instrument server.

To facilitate the addition of functionality which would result in more
"parts" in the future, `into_parts` returns a non-exhaustive struct,
rather than a tuple of parts.

Closes: #428

hawkw commented on console-subscriber/src/lib.rs:
```diff
@@ -957,9 +957,110 @@ impl Server {
+
+    /// The aggregate handle.
+    ///
+    /// See the documentation for [`AggregatorHandle`] for details.
```
Since dropping this shuts down the aggregator task, we should probably add a Big Scary Warning on this field making it clear that it must not be dropped.

hds replied:
I've removed the drop behaviour as you suggested in a later comment, so I think this doesn't apply any more.

hawkw commented on console-subscriber/src/lib.rs:
```diff
@@ -957,9 +957,110 @@ impl Server {
+    /// See the documentation for [`InstrumentServer`] for details.
+    pub instrument_server: InstrumentServer<Server>,
+
+    /// The aggregate handle.
```
nit: how about
```suggestion
    /// A handle to the background worker task responsible for aggregating trace data.
```

hawkw commented on console-subscriber/src/lib.rs:
```diff
@@ -957,9 +957,110 @@ impl Server {
+    pub aggregator_handle: AggregatorHandle,
+}
+
+/// Aggregator handle.
```
This documentation doesn't really explain to a user what the heck an "aggregator handle" _is_ or why they should care. Let's make sure this actually states what it does, and what it's for.

hds replied:
I've tried to add more context here.

hawkw commented on console-subscriber/src/lib.rs:
```diff
@@ -957,9 +957,110 @@ impl Server {
+    /// worker running.
+    ///
+    /// Note that a server spawned in this way will overwrite any value set by
+    /// [`Builder::server_addr`] as the user becomes responsible for defining
```
```suggestion
    /// [`Builder::server_addr`], as the user becomes responsible for defining
```

hawkw commented on console-subscriber/src/lib.rs:
```diff
@@ -957,9 +957,110 @@ impl Server {
+    /// Returns the parts needed to spawn a gRPC server and keep the aggregation
+    /// worker running.
+    ///
+    /// Note that a server spawned in this way will overwrite any value set by
```
```suggestion
    /// Note that a server spawned in this way will disregard any value set by
```

hawkw commented on console-subscriber/src/lib.rs:
```diff
@@ -957,9 +957,110 @@ impl Server {
+impl Drop for AggregatorHandle {
+    fn drop(&mut self) {
+        self.join_handle.abort();
+    }
```
I'm not convinced we should always do this. Instead, perhaps, the `ServerParts` should just give the user access to the `JoinHandle` so they can decide whether they want to abort the task? The behavior  of `serve_with` was specifically intended to ensure that the aggregator is shut down if the server task ends.

hds replied:
@hawkw OK, that makes sense. I was hoping to avoid abstract away the fact that this is a join handle underneath. Do you think there's any value in that?

I see 2 options here:
* Keep the `AggregatorHandle` without the `impl Drop` and add an `abort()` method to explicitly shut down the aggregator.
* Expose the join handle as is.

Do you have a specific preference?

hds replied:
@hawkw i went with the first option. Let me know what you think.

hawkw replied:
> OK, that makes sense. I was hoping to avoid abstract away the fact that this is a join handle underneath

Hmm...now that I think about it a bit more, wouldn't the ideal lower-level interface just be to expose the future itself, instead of spawning it and handing the user a `JoinHandle`? Then, the user could spawn the aggregator task wherever they want to, which could be useful if the user is using a non-Tokio runtime, or has multiple Tokio runtime, and wants to control where the aggregator task is spawned. Also, we can avoid having to think about the `JoinHandle` entirely if it's up to the user to spawn the aggregator. 

What do you think?

hds replied:
I think that makes a lot of sense.

Unfortunately I wasn't able to find a way to return and then later use a future (dyn, boxed, both, or otherwise), so I've made the Aggregator public with just the `run()` method which can be used to spawn it.

How does that look?

hds committed:
Apply suggestions from code review

Co-authored-by: Eliza Weisman <eliza@buoyant.io>

hds committed:
make `AggregatorHandle` abort explcit

Instead of aborting the aggregator task upon dropping the
`AggregatorHandle`, we now provide an explicit `abort` method. If the
user chooses, they can discard the aggregator handle and while they will
lose the possibility to clean up the aggregator task, they will be
otherwise unaffected.

hds committed:
fixed doc link

hds committed:
expose the Aggregator instead of a handle to abort it

As discussed in code review, for such a low level api as
`Server::into_parts`, it makes sense to allow the user to spawn the
aggregator where they like, rather than spawning it internally.

Since I couldn't find a way to return and use a `dyn Future` (boxed or
otherwise), the `Aggregator` has been made public with a (async) single
function `run()` which will start it's run loop.

hawkw reviewed (approved):
Thanks for going along with my last-minute suggested changes, this looks great!

hawkw committed:
Merge branch 'main' into hds/subscriber-server-into-parts
